### PR TITLE
feat(parachain): Implement request and response message for /req_statement/1 protocol

### LINKED ISSUE #3343: Implement request and response message for `/req_statement/1` protocol
## Issue summary
- Should be done after https://github.com/ChainSafe/gossamer/pull/3334/files is merged
- Request struct must implement `network.Message` interface
- Response struct must implement `network.ResponseMessage`

## Other information and links

- https://spec.polkadot.network/chapter-anv#net-msg-statement-fetching-request
- https://github.com/paritytech/polkadot/blob/47c8c36d6e3f90da77ebccb8b672e4fb4f33c84c/node/network/protocol/src/request_response/outgoing.rs#L40
- https://spec.polkadot.network/chapter-anv#net-msg-statement-fetching-response

### PR DESCRIPTION
**Pull request by axaysagathiya**
## Changes

- Added `StatementFetchingRequest` and `StatementFetchingResponse`  varying data types.

- I have implemented '[network.Message](https://github.com/ChainSafe/gossamer/blob/457fe39f7a12cb9eb04de5b6a315d38825247d68/dot/network/message.go#L30)' interface in `StatementFetchingRequest` and '[network.ResponseMessage](https://github.com/ChainSafe/gossamer/pull/3334/files#diff-5794e04e36068d98b7c6d2280c7e0b41fecca0b90bd6dd4576cd62b951ab587aR88)' interface in `StatementFetchingResponse` as they will be passed into [this](https://github.com/ChainSafe/gossamer/pull/3334/files#diff-5794e04e36068d98b7c6d2280c7e0b41fecca0b90bd6dd4576cd62b951ab587aR32) function as `req` and `res`.

- I didn't want to create a new YAML file here. so I decided to rename the YAML file name and variable(in which data of the YAML file getting unmarshalled) name so that I can use them in this PR.

## Tests

```sh
go test -tags integration github.com/ChainSafe/gossamer
```

## Issues

Fixes #3343

## Primary Reviewer

@kishansagathiya @kanishkatn @edwardmack

### CODE REVIEW

File: lib/parachain/statement_fetching.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+type StatementFetchingRequest struct {
+	// Data needed to locate and identify the needed statement.
+	RelayParent common.Hash `scale:"1"`
+
+	// Hash of candidate that was used create the `CommitedCandidateRecept`.
+	CandidateHash CandidateHash `scale:"2"`
+}
+
+// Encode returns the SCALE encoding of the StatementFetchingRequest.
+func (s *StatementFetchingRequest) Encode() ([]byte, error) {
```
Comment (jimjbrettj) [inline lib/parachain/statement_fetching.go:20]: Same feedback here, I dont think these are necessary are they?

**Review by jimjbrettj — COMMENTED** [review]
Looks good to me, just same question as other PR about wrapping scale

File: lib/parachain/statement_fetching.go (hunk 1 above)
Comment (axaysagathiya) [reply lib/parachain/statement_fetching.go:20]: yup,
I have implemented the '[network.Message](https://github.com/ChainSafe/gossamer/blob/457fe39f7a12cb9eb04de5b6a315d38825247d68/dot/network/message.go#L30)' interface as 'StatementFetchingRequest' will be passed into [this](https://github.com/ChainSafe/gossamer/pull/3334/files#diff-5794e04e36068d98b7c6d2280c7e0b41fecca0b90bd6dd4576cd62b951ab587aR32) function as 'req'.

**Review by kanishkatn — COMMENTED** [review]
I feel the changes to the modified files should be in a new PR since they are unrelated. Looks good to me otherwise!

**Comment by axaysagathiya** [comment]
@kanishkatn Can you be more specific about what should I move to separate PR?

**Comment by kanishkatn** [comment]
> @kanishkatn Can you be more specific about what should I move to separate PR?

haha sure. This PR contains 2 addtions, 2 modifications and a rename of a file. My comment was talking about the modified files. 

I think I had a similar conversation in another PR as well. I am okay with it for now, but this is generally considered bad practice. Curious to know what rest of the team thinks!

**Comment by axaysagathiya** [comment]
@kanishkatn okay.
I didn't want to create a new YAML file here. so I decided to rename the file name and variable name so that I can use them in this PR.

**Comment by kanishkatn** [comment]
> @kanishkatn okay. I didn't want to create a new YAML file here. so I decided to rename the file name and variable name so that I can use them in this PR.

I see. I guess you could just update the description of the PR!

**Review by kanishkatn — APPROVED** [review]

**Review by EclesioMeloJunior — APPROVED** [review]

File: lib/parachain/statement_fetching_test.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+				RelayParent:   getDummyHash(4),
+				CandidateHash: CandidateHash{Value: getDummyHash(4)},
+			},
+			expectedEncode: common.MustHexToBytes(testDataStatement["all4InCommonHash"]),
+		},
+		{
+			name: "all_7_in_hash",
+			request: StatementFetchingRequest{
+				RelayParent:   getDummyHash(7),
+				CandidateHash: CandidateHash{Value: getDummyHash(7)},
```
Comment (kishansagathiya) [inline lib/parachain/statement_fetching_test.go:42]: Let's add one more example with a more realistic hash values.

File: lib/parachain/statement_fetching_test.go [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+			require.NoError(t, err)
+			require.Equal(t, c.expectedEncode, actualEncode)
+		})
+	}
+}
+
+func TestStatementFetchingResponse(t *testing.T) {
+	t.Parallel()
+
+	hash5 := getDummyHash(5)
```
Comment (kishansagathiya) [inline lib/parachain/statement_fetching_test.go:62]: try with a more realistic hash value

**Review by kishansagathiya — APPROVED** [review]
Looks good!

**Pull request merged**